javascript/react-js/react-testing-part-one.md

```diff
@@ -0,0 +1,177 @@
+        Click Me
+      </button>
+      <h1>{heading}</h1>
+    </>
+  );
+};
+
+export default App;
+~~~
+
+Let's test if the button works as intended:
+
+~~~javascript
+// App.test.js
+
+import React from "react";
+import { render, fireEvent } from "@testing-library/react";
+import App from "./App";
+
+test("magnificent monkeys renders", () => {
+  const { container } = render(<App />);
+  expect(container).toMatchSnapshot();
+});
+
+test("radical rhinos renders after button click", () => {
+  const { getByRole } = render(<App />);
+  const button = getByRole("button");
+
+  fireEvent.click(button);
+
+  expect(getByRole("heading").textContent).toMatch(/radical rhinos/i);
+});
+~~~
+
+The tests speak for themselves. In the first test, we utilize snapshots to check whether all the nodes render as we expect them to. In the second test, we simulate a click event. Then we check if the heading changed. `toMatch` is one of the various assertions we could have made. 
+
+It's also important to note that after every test, React Testing Library unmounts the rendered components. That's why we render for each test. For a lot of tests for a component, the `beforeEach` jest function could prove handy.
+
+### What are Snapshots?
+
+Snapshot testing is just comparing our rendered component with an associated snapshot file. For example, the snapshot file which was automatically generated after we ran the _"magnificent monkeys renders"_ test was:
+
+~~~javascript
+// App.test.js.snap
+
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`magnificent monkeys render 1`] = `
+<div>
+  <button
+    type="button"
+  >
+    Click Me
+  </button>
+  <h1>
+    Magnificent Monkeys
+  </h1>
+</div>
+`;
+~~~
+
+It's an HTML representation of the `App` component. And it will be compared against the `App` in future snapshot assertions. If the `App` changes even slightly, the test fails. 
+
+Snapshot tests are fast and easy to write. One assertion saves us from writing multiple lines of code. For example, with a `toMatchSnapshot`, we're spared of asserting the existence of the button and the heading. They also don't let unexpected changes creep into our code. Read all about what can be achieved with snapshots [here](https://jestjs.io/docs/snapshot-testing)
+
+Snapshots might seem the best thing that has happened to us while testing thus far. But we are forced to wonder, _what_ exactly are we testing? What's being validated? If a snapshot passes, what does it convey about the correctness of the component? 
+
+The other issue with snapshots is false negatives. Even the insignificant of changes compels the test to fail. Fixing punctuation? Snapshot will fail. Replacing an HTML tag to a more semantic one? Snapshot will fail. This might cause us to lose our confidence in the test suite altogether. Snapshots aren't inherently bad; they do serve a purpose. But it's beneficial to understand when to snapshot, and when not to snapshot. 
```

Comment:
```suggestion
The other issue with snapshots is false negatives. Even the most insignificant of changes compel the test to fail. Fixing punctuation? Snapshot will fail. Replacing an HTML tag to a more semantic one? Snapshot will fail. This might cause us to lose our confidence in the test suite altogether. Snapshots aren't inherently bad; they do serve a purpose. But it's beneficial to understand when to snapshot, and when not to snapshot. 
```